hphp/hsl/src/str/introspect_l.php

```diff
@@ -0,0 +1,296 @@
+  int $offset = 0,
+)[]: bool {
+  if ($needle === '') {
+    if ($offset === 0) {
+      return true;
+    }
+    $length = length_l($locale, $haystack);
+    if ($offset > $length || $offset < -$length) {
+      throw new \InvalidArgumentException(
+        format('Offset %d out of bounds for length %d', $offset, $length)
+      );
+    }
+    return true;
+  }
+  return search_l($locale, $haystack, $needle, $offset) !== null;
+}
+
+/**
+ * Returns whether the "haystack" string contains the "needle" string
+ * (case-insensitive).
+ *
+ * An optional offset determines where in the haystack the search begins. If the
+ * offset is negative, the search will begin that many characters from the end
+ * of the string. If the offset is out-of-bounds, a ViolationException will be
+ * thrown.
+ *
+ * Locale-specific rules for case-insensitive comparisons will be used, and
+ * strings will be normalized before comparing if the locale specifies an
+ * encoding that supports multiple representations of the same characters, such
+ * as UTF-8.
+ *
+ * - To search for the needle case-sensitively, see `Str\contains_l()`.
+ * - To get the position of the needle case-insensitively, see `Str\search_ci_l()`.
+ */
+function contains_ci_l(
+  Locale\Locale $locale,
+  string $haystack,
+  string $needle,
+  int $offset = 0,
+)[]: bool {
+  if ($needle === '') {
+    if ($offset === 0) {
+      return true;
+    }
+    $length = length_l($locale, $haystack);
+    if ($offset > $length || $offset < -$length) {
+      throw new \InvalidArgumentException(
+        format('Offset %d out of bounds for length %d', $offset, $length)
+      );
+    }
+    return true;
+  }
+  return search_ci_l($locale, $haystack, $needle, $offset) !== null;
+}
+
+/**
+ * Returns whether the string ends with the given suffix.
+ *
+ * Locale-specific rules for case-insensitive comparisons will be used, and
+ * strings will be normalized before comparing if the locale specifies an
+ * encoding that supports multiple representations of the same characters, such
+ * as UTF-8.
+ *
+ * For a case-insensitive check, see `Str\ends_with_ci_l()`.
+ */
+function ends_with_l(
+  Locale\Locale $locale,
+  string $string,
+  string $suffix,
+)[]: bool {
+  /* HH_FIXME[4390] missing [] */
+  return _Str\ends_with_l($string, $suffix, $locale);
+}
+
+/**
+ * Returns whether the string ends with the given suffix (case-insensitive).
+ *
+ * Locale-specific rules for case-insensitive comparisons will be used, and
+ * strings will be normalized before comparing if the locale specifies an
+ * encoding that supports multiple representations of the same characters, such
+ * as UTF-8.
+ *
+ * For a case-sensitive check, see `Str\ends_with_l()`.
+ */
+function ends_with_ci_l(
+  Locale\Locale $locale,
+  string $string,
+  string $suffix,
+)[]: bool {
+  /* HH_FIXME[4390] missing [] */
+  return _Str\ends_with_ci_l($string, $suffix, $locale);
+}
+
+/**
+ * Returns the length of the given string in characters.
```

Comment:
Something that might be important to note here, `Str\length()` is O(1).
Is `Str\length_l()` `O(n)` for every call?

Reply:
That depends on the locale; for UTF-8 yes, but single-byte locales, no.

UTF-16 would also be O(n), while UCS-2 would be O(1)